### PR TITLE
YD-499 Create first device during migration

### PR DESCRIPTION
**Pull request by Bert-R**
Along with this corrected an earlier bug (user and device anonymized weren't linked).

This functionality is covered with Spring migration tests. Considerably enhanced our Spring JUnit integration test utilities.

### CODE REVIEW

**Comment by yonabuild** [comment]
Build triggered. sha1 is merged.

**Comment by yonabuild** [comment]
Build started sha1 is merged.

**Comment by Bert-R** [comment]
There is still a TODO in ``BuddyDeviceChangeMessageDtoTest``, to add unit tests to verify update and delete of devices on the buddy side. This will be done in parallel with the first review round.

**Comment by yonabuild** [comment]
Build finished.

**Comment by yonabuild** [comment]
Build triggered. sha1 is merged.

File: core/src/main/java/nu/yona/server/subscriptions/entities/BuddyDeviceChangeMessage.java [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+import nu.yona.server.crypto.seckey.SecretKeyUtil;
+import nu.yona.server.device.service.DeviceChange;
+import nu.yona.server.messaging.entities.BuddyMessage;
+
+@Entity
+public class BuddyDeviceChangeMessage extends BuddyMessage
+{
+	@Transient
+	private DeviceChange change;
+	private byte[] changeCiphertext;
```
Comment (aristotelos) [inline core/src/main/java/nu/yona/server/subscriptions/entities/BuddyDeviceChangeMessage.java]: Do we really have to encrypt the change? Seems superfluous

File: core/src/main/java/nu/yona/server/subscriptions/service/BuddyService.java [hunk 2]
```diff
@@ -706,4 +707,46 @@ public UserDto getUserOfBuddy(UUID userId, UUID buddyUserId)
 				.orElseThrow(() -> BuddyNotFoundException.notFoundForUser(userId, buddyUserId));
 		return UserDto.createInstanceWithBuddyData(userEntity, BuddyUserPrivateDataDto.createInstance(buddy));
 	}
+
+	@Transactional
+	public void processDeviceChange(UUID idOfRequestingUser, UUID relatedUserAnonymizedId, DeviceChange change,
+			UUID deviceAnonymizedId, Optional<String> oldName, Optional<String> newName)
+	{
+		User user = userService.getValidatedUserbyId(idOfRequestingUser);
```
Comment (aristotelos) [inline core/src/main/java/nu/yona/server/subscriptions/service/BuddyService.java]: `getValidatedUserById`

File: core/src/main/java/nu/yona/server/subscriptions/service/BuddyService.java [hunk 3]
```diff
@@ -706,4 +707,46 @@ public UserDto getUserOfBuddy(UUID userId, UUID buddyUserId)
...
+				removeDeviceFromBuddy(buddy, deviceAnonymizedId);
+				break;
+			default:
+				throw new IllegalArgumentException("Change " + change + " is unknown");
+		}
+	}
+
+	private void addDeviceToBuddy(Buddy buddy, UUID deviceAnonymizedId, String name)
+	{
+		BuddyDevice device = BuddyDevice.createInstance(name, deviceAnonymizedId, true); // Assume VPN is connected
```
Comment (aristotelos) [inline core/src/main/java/nu/yona/server/subscriptions/service/BuddyService.java]: Couldn't this assumption lead to problems?

File: core/src/main/resources/messages/messages.properties [hunk 4]
```diff
@@ -8,6 +8,9 @@
 message.user.account.deleted=User account was deleted
 message.user.removed.buddy=User removed you as a buddy
 message.buddy.user.info.changed=User changed personal info
+message.buddy.device.added=User added a device named {0}
```
Comment (aristotelos) [inline core/src/main/resources/messages/messages.properties]: Maybe better to use quotes?

File: core/src/test/java/nu/yona/server/subscriptions/service/PrivateUserDataMigrationServiceIntegrationTest.java [hunk 5]
```diff
@@ -85,95 +95,75 @@ public void upgrade(User userEntity)
...
-	public void teardownForAll()
-	{
-		PrivateUserDataMigrationService.setMigrationSteps(originalMigrationSteps);
 	}
 
 	@Before
 	public void setUpPerTest()
 	{
+		userAnonymizedRepository.deleteAll();
 		JUnitUtil.setUpRepositoryMock(mockUserRepository);
```
Comment (aristotelos) [inline core/src/test/java/nu/yona/server/subscriptions/service/PrivateUserDataMigrationServiceIntegrationTest.java]: Why not convert those mocks too?

File: core/src/test/java/nu/yona/server/test/util/JUnitUtil.java [hunk 6]
```diff
@@ -65,4 +71,43 @@ public Object answer(InvocationOnMock invocation) throws Throwable
...
+		MessageSource anonymousMessageSource = MessageSource.createInstance();
+		UserAnonymized johnAnonymized = UserAnonymized.createInstance(anonymousMessageSource.getDestination(),
+				Collections.emptySet());
+		UserAnonymized.getRepository().save(johnAnonymized);
+
+		byte[] initializationVector = CryptoSession.getCurrent().generateInitializationVector();
+		MessageSource namedMessageSource = MessageSource.createInstance();
+		UserPrivate userPrivate = UserPrivate.createInstance(nickname, vpnPassword, johnAnonymized.getId(),
+				anonymousMessageSource.getId(), namedMessageSource);
+		return new User(UUID.randomUUID(), initializationVector, firstName, lastName, mobileNumber, userPrivate,
```
Comment (aristotelos) [inline core/src/test/java/nu/yona/server/test/util/JUnitUtil.java]: Shouldn't this be saved in the `UserRepository` too?

File: core/src/main/java/nu/yona/server/device/service/DeviceServiceException.java [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+public class DeviceServiceException extends YonaException
+{
+	private static final long serialVersionUID = 8798601529916186731L;
+
+	private DeviceServiceException(String messageId, Serializable... parameters)
+	{
+		super(messageId, parameters);
+	}
+
+	protected DeviceServiceException(HttpStatus statusCode, String messageId, Serializable... parameters)
```
[the remark this answers is not in the document]
Comment (Bert-R) [reply core/src/main/java/nu/yona/server/device/service/DeviceServiceException.java]: Copy/paste error. Corrected it.

File: core/src/main/java/nu/yona/server/subscriptions/entities/BuddyDeviceChangeMessage.java (hunk 1 above)
Comment (Bert-R) [reply core/src/main/java/nu/yona/server/subscriptions/entities/BuddyDeviceChangeMessage.java]: I agree. Changed it.

File: core/src/main/java/nu/yona/server/subscriptions/service/BuddyService.java (hunk 2 above)
Comment (Bert-R) [reply core/src/main/java/nu/yona/server/subscriptions/service/BuddyService.java]: Done.

File: core/src/main/resources/messages/messages.properties (hunk 4 above)
Comment (Bert-R) [reply core/src/main/resources/messages/messages.properties]: OK

File: core/src/test/java/nu/yona/server/subscriptions/service/PrivateUserDataMigrationServiceIntegrationTest.java (hunk 5 above)
Comment (Bert-R) [reply core/src/test/java/nu/yona/server/subscriptions/service/PrivateUserDataMigrationServiceIntegrationTest.java]: My approach is that I use a Mockito mock wherever I can and only use our own mocks when Mockito is insufficient.

File: core/src/test/java/nu/yona/server/test/util/JUnitUtil.java (hunk 6 above)
Comment (Bert-R) [reply core/src/test/java/nu/yona/server/test/util/JUnitUtil.java]: OK, will do.

File: core/src/main/java/nu/yona/server/subscriptions/service/BuddyService.java (hunk 3 above)
Comment (Bert-R) [reply core/src/main/java/nu/yona/server/subscriptions/service/BuddyService.java]: Let's address this when handling the VPN status issue. See [YD-514](http://jira.yona.nu/browse/YD-514).

**Comment by yonabuild** [comment]
Build triggered. sha1 is merged.

**Comment by yonabuild** [comment]
Build started sha1 is merged.

**Comment by Bert-R** [comment]
Pushed an update that addresses the review comments.

**Comment by yonabuild** [comment]
Build finished.

**Comment by yonabuild** [comment]
Build triggered. sha1 is merged.

**Comment by yonabuild** [comment]
Build started sha1 is merged.

**Comment by yonabuild** [comment]
Build finished.

File: core/src/test/java/nu/yona/server/subscriptions/service/PrivateUserDataMigrationServiceIntegrationTest.java [hunk 8]
```diff
@@ -115,32 +90,24 @@ public void upgrade(User userEntity)
...
 			richard = JUnitUtil.createRichard();
 		}
 		reset(userRepository);
 	}
 
+	@Override
+	protected Map<Class<?>, Repository<?, ?>> getRepositories()
+	{
+		Map<Class<?>, Repository<?, ?>> repositoriesMap = new HashMap<>();
+		repositoriesMap.put(MessageSource.class, mockMessageSourceRepository);
```
Comment (aristotelos) [inline core/src/test/java/nu/yona/server/subscriptions/service/PrivateUserDataMigrationServiceIntegrationTest.java]: Why are the user repositories not included here as they are in `BuddyDeviceChangeMessageDtoTest`?

Comment (Bert-R) [reply core/src/test/java/nu/yona/server/subscriptions/service/PrivateUserDataMigrationServiceIntegrationTest.java]: It's the opposite: why are they included in ``BuddyDeviceChangeMessageDtoTest`` and ``AddFirstDeviceTest``? And the answer to that is: for no good reason.

I'll push an update that removes them there.

**Comment by yonabuild** [comment]
Build triggered. sha1 is merged.

**Comment by yonabuild** [comment]
Build started sha1 is merged.

**Comment by Bert-R** [comment]
Given that the change set was approved and that I addressed the final question, I have merged this PR, so I can build further on it.

**Comment by aristotelos** [comment]
Thanks, I planned to do that myself but got no time yet :)

**Comment by Bert-R** [comment]
No problem at all.

**Pull request merged**